rbac/api/cross_access/util.py

```diff
@@ -0,0 +1,53 @@
+#
+# Copyright 2019 Red Hat, Inc.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+#
+
+"""Handler for cross-account request clean up."""
+import logging
+
+from django.db.models import Q
+from django.utils import timezone
+from prometheus_client import Summary
+from tenant_schemas.utils import tenant_context
+
+from api.models import CrossAccountRequest, Tenant
+
+logger = logging.getLogger(__name__)  # pylint: disable=invalid-name
+
+# Create processing time metric
+PROCESSING_TIME = Summary(
+    "cross_account_expiry_processing_second", "Time spent checking and expiring cross-account requests"
+)
+
+
+@PROCESSING_TIME.time()
+def check_cross_request_expiry():
+    """Check if a cross-account requests have expired, tag them if so."""
+    with tenant_context(Tenant.objects.get(schema_name="public")):
+        expired_cars = []
+        cars = CrossAccountRequest.objects.filter(Q(status="pending") | Q(status="approved"))
+        logger.info("Running expiry check on %d cross-account requests.", len(cars))
+        for car in cars:
+            logger.debug("Checking for expiration of cross-account request %s.", car.pk)
+            if car.end_date < timezone.now() or (car.status == "pending" and car.start_date < timezone.now()):
+                logger.info("Expiring cross-account request with uuid: %s", car.pk)
+                car.status = "expired"
+                expired_cars.append(car.pk)
+                car.save()
```

Comment:
A discussion here:
In the plan, the created principal have to be removed for approved cars. But that's when we planned principals have groups, policies roles attached to them. Should we keep them? seems no harm though.

Reply:
Ah! I thought I'd responded to this earlier, but I think you're right, we should pull it out. I can get that added in real quick.

Reply:
👍Thanks! And this actually reminds me missing sth😅 https://github.com/RedHatInsights/insights-rbac/pull/439

Reply:
Also, would it be a concern if a principal has multiple cars, and only one has expired?

Reply:
Does that mean `remove_cross_principal` goes away based on this conversation?

Reply:
Ah, good point! I thought there would be only one approved request in one account for one associate, but seems we do support multiple requests. Maybe we could check if there's any approved left before removing them?

Reply:
Or maybe, check that all are "expired"? If not, leave it? Also, historically would we want to keep referential integrity for audit purposes, similar to how we aren't going to actually delete the cars?

Reply:
@wcmitchell - @astrozzc and I were discussing more in the dev sync, and it seems like we may just want to leave the principals for now, for historical reference and to ensure we don't remove a principal who still has CARs tied to it.

Reply:
That seems reasonable. We'd still need to remove the roles from the principal, I guess? Or...hrm. We'll need to do something to actually shut down the access, right?